gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs

```diff
@@ -179,6 +183,13 @@ runUI ma = do
   takeMVar r
 
 runUI_ :: IO () -> IO ()
-runUI_ ma = void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $ do
-  ma
-  return False
+runUI_ ma = do
+  tId <- myThreadId
+
+  void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $ do
+    -- Any exception in the gtk ui thread will be rethrow in the
+    -- calling thread.
+    -- This ensure that this exception won't terminate the application without any control.
+    -- HOWEVER, it may be raised ANYWHERE in the calling thread.
```

Comment:
The calling thread will always be in `runUI` waiting for `takeMVar` to complete (or _about_ to call `takeMVar`) - right? So there is no question about where the exception will be thrown to - right?

Anyway, this looks good to my (limited) understanding. I suppose an alternative would be to change `runUI` so it catches `SomeException` with `try @SomeException ma >>= putMVar r` and then unwraps the `Either SomeException a` and re-`throwIO`s the `SomeException`. I don't know if that would be better or worse.

Reply:
@Dretch my understanding is that `Gdk.threadsAddIdle _ foo` will immediately terminate. It only schedules`foo` to be executed later in the gtk thread. So `runUI_` will terminate immediately. So technically, `runUI`, from my point of view, is a non blocking action and I don't really understand the rational for the `takeMVar` / `putMVar`.

Later, in the gtk main loop, the `IO` action will be called and my raise an exception in the thread which was running `runUI`.

I do admit that I did not really tried to fully understand the logic and perhaps it can be refactored differetly.

Reply:
Hmm, `runUI_` is non-blocking, but `runUI` is blocking. It needs to be because it returns the result of the callback. `takeMVar` / `putMVar` are used to wait for the result to be produced and to pull it from the other thread.

Reply:
Haaa, yes. Sorry, you are right. Yes, the logic is perfectly fine and you are right, the exception will be raised inside `runUI`. Sorry.

Reply:
(concurrent programming is hard ;)